Better template caching for container environments

kloon opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:
This PR replaces #23783 It works based on the suggestion by @markjaquith where we remove ABSPATH from the template location before caching it and then adding it back right before loading it.

Closes #23751

### How to test the changes in this Pull Request:

1. See #23783 for instructions on testing

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Fix - Better template caching support for multi-server environments.

kloon committed:
Store template cache value without ABSPATH to avoid issues with multi container environments.

kloon committed:
Cater for blank paths, do not prepend ABSPATH when template path blank.

kloon commented:
Moving this to In Progress, since this PR seems to break the unit tests. Will have to dig a bit to find out why since browsing the site it works fine.

carl-alberto commented:
Thanks for the progress and update with this @kloon ! :)

peterfabian commented on includes/wc-core-functions.php:
```diff
@@ -185,7 +185,16 @@ function wc_get_template_part( $slug, $name = '' ) {
 
-		wp_cache_set( $cache_key, $template, 'woocommerce' );
+		// Remove ABSPATH from template cache to avoid issues with multi container having different paths.
+		if ( 0 === strpos( $template, ABSPATH ) ) {
```
Perhaps using `WC_ABSPATH` could be considered to account for cases where WooCommerce is not running under `ABSPATH`, as in the context of our unit test. That should provide the same behaviour as previously (so hopefully not break bw compat) and shouldn't be a problem when `ABSPATH`==`WC_ABSPATH`?

wss-chadical replied:
Why check for the strpos... ? str_replace has to check anyways

peterfabian reviewed (commented):
Please see some of my suggestions...

JulienMelissas commented:
I'm currently at #WCEU so unable to look at this right now but will try and take a look before my vacation on Monday! Thanks for your work.

Uriel1339 commented:
Since 3.6.5 just got released will this make it into 3.6.6? It is still quite a severe issue for all Pantheon users and especially with recent security patches we cannot wait much longer for this issue to be resolved.

JulienMelissas commented:
@Uriel1339 -- in case you missed it, my fix is working for the time being. Here is a link: https://gist.github.com/JulienMelissas/d3601af950221432a109c8c4b1617b87

I'd rather this feature get tested properly instead of rushing to have it in a release as soon as possible, have you been able to test this PR?

Uriel1339 commented:
> @Uriel1339 -- in case you missed it, my fix is working for the time being. Here is a link: https://gist.github.com/JulienMelissas/d3601af950221432a109c8c4b1617b87
> 
> I'd rather this feature get tested properly instead of rushing to have it in a release as soon as possible, have you been able to test this PR?

My intention is not to rush but rather to check the status, since the last update was quite some time ago. I don't have the means to reproduce an environment similar to Pantheons and I'm not risking our company's flagship site of whether or not it *may* work. 

I was hoping that @carl-alberto might be able to test your integration, considering he has the resources for testing and was also the individual who originally helped me out over at Pantheon tech support.

JulienMelissas commented:
Gotcha. Well we are on Pantheon (6-10 containers running) and do hundreds
of orders a day with this as an mu-plugin in production, so it’s def. been
tested. Pretty sure Pantheon support has pointed a few other customers to
my gist to help for the time being.

On Wed, Jul 10, 2019 at 10:18 PM Andreas Lopez <notifications@github.com>
wrote:

> @Uriel1339 <https://github.com/Uriel1339> -- in case you missed it, my
> fix is working for the time being. Here is a link:
> https://gist.github.com/JulienMelissas/d3601af950221432a109c8c4b1617b87
>
> I'd rather this feature get tested properly instead of rushing to have it
> in a release as soon as possible, have you been able to test this PR?
>
> My intention is not to rush but rather to check the status, since the last
> update was quite some time ago. I don't have the means to reproduce an
> environment similar to Pantheons and I'm not risking our company's flagship
> site of whether or not it *may* work.
>
> I was hoping that @carl-alberto <https://github.com/carl-alberto> might
> be able to test your integration, considering he has the resources for
> testing and was also the individual who originally helped me out over at
> Pantheon tech support.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/woocommerce/woocommerce/pull/23971?email_source=notifications&email_token=AARMGTNKDDPT23Y4J7IQIP3P62J6FA5CNFSM4HZSVJYKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZVJFZQ#issuecomment-510300902>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AARMGTNO4PL7FIWW46PNTWLP62J6FANCNFSM4HZSVJYA>
> .
>
-- 
- Julien

peterfabian commented:
At this point, we're not completely sure if there will be a 3.6.6, as we're near tagging beta for 3.7. As this change needed a bit more testing, we skipped it for 3.6.5. I'd expect it to land either in 3.7 or the next fix release. 

It would be ideal if you gentlemen could help us test this PR on your staging/preprod environments as well, to cover more possible configurations.

Uriel1339 commented:
We just implemented the fix from @JulienMelissas on www.jnequipment.com and it appears to have worked. We are running 3.7. Without Juliens fix our website completely breaks with the aforementioned issues above. With the PHP script running, everything works as intended.

I think it can be merged into WooCommerce.

JulienMelissas commented:
@Uriel1339, glad to hear my temporary solution is working for you. That said, this PR doesn't have the same code as my temp workaround.

Does anyone have a status update on this ticket? I was holding off to test it until @peterfabian's suggestions above had been implemented or at least discussed on.

Happy to test whatever, just need to know when it's ready to test. Thanks!

wss-chadical commented:
We are a composer based wordpress running on Pantheon.

One key thing in this PR that is not being considered is that in composer based WP installations the ABSPATH and WP_CONTENT paths can be different.
For example our paths look like this
ABSPATH: /app/web/wp
WP_CONTENT: /app/web/wp-content

So, themes are not under ABSPATH, they are under WP_CONTENT.

We were able to workaround this by setting a DOCROOT variable which is based on the location of wp-config.php... 
DOCROOT: /app/web

Uriel1339 commented:
Still have @JulienMelissas implication built-in as without it our site won't run WooCommerce. Is there any more pushing done towards getting this into the core?

JulienMelissas commented:
Looks like this has been removed from the 3.9 milestone? I would love to hear from the WC team on what we (the folks who are experiencing issues with this) can do to help? I'm a little confused on the status of this ticket as I mentioned above, so just wanted to make sure I can use my time valuably towards helping to get this resolved. Thanks!

kfoubert commented:
I would like to second the comment made by @JulienMelissas . Our company started the original ticket when we ran into a number of issues with hosting on Patheon.

wss-chadical commented:
Yes please.

Uriel1339 commented:
I fourth that comment by @JulienMelissas . Like @kfoubert I'm hosting with Pantheon.

frederickalcantara commented:
I am also the fifth to concur about the issues that @JulienMelissas and the rest are expressing. I am also having issues with my Pantheon.

carl-alberto commented:
I would also like to 6th this if that is the case! Been watching the progress for the past months but hands full at the moment to do a PR if none is implemented yet

jasoncam commented:
7th on this, hosting w Pantheon

peterfabian commented:
Thanks for expressing your interest here. This has been previously deprioritized as other projects came in. However, seeing it has a lot of interest, I've pushed it up our backlog to give it a higher priority. It won't make it into 3.9, but we'll try to get it into the next release.

JulienMelissas commented:
@peterfabian thank you for the update, this is great news! This would be really helpful to have implemented for us.

My question is still unanswered from the WC team so I will state it again here:
>I would love to hear from the WC team on what we (the folks who are experiencing issues with this) can do to help?

Is there anything we can do to help your team get this tested or moving? Would it be helpful for @carl-alberto or someone from Pantheon to set up a site on Pantheon with multiple app servers for your team so you can use it as an example? Do you need help testing the solution in the current PR? I would love to help with this PR, or another one, but don't know enough about the WC team's process to be helpful at this point. Thank you for any clarification you can provide.

ObliviousHarmony committed:
Merge branch 'master' into refactor/23783

kloon commented:
@JulienMelissas we have asked prior for everyone experience this issue to please help test but have not gotten any response yet.

We would love if there are Pantheon users that would be willing to test an this for us, we do not have access to Pantheon hosting at present but we will be sure to include them in our hosting testing we do before each release in the near future. cc @juliaamosova

Uriel1339 commented:
@kloon what kind of testing do you exactly need done? We are running woocommerce on Pantheon.

kfoubert commented:
This was originally requested by someone at Pantheon. I'll try to find the name, but they should be the ones that could help. This also doesn't affect just Pantheon hosting, it also affects anyone who happens to have load balancing and load balancing is not setup to keep a person's session on one server.

wss-chadical commented:
> This was originally requested by someone at Pantheon. I'll try to find the name, but they should be the ones that could help. This also doesn't affect just Pantheon hosting, it also affects anyone who happens to have load balancing and load balancing is not setup to keep a person's session on one server.

Agreed - this is not really "pantheon" issue... any wp instance running woocommerce on multiple appservers using a shared wp cache storage backend such as redis.

JulienMelissas commented:
@kloon - thanks for the reply. I'm happy to test the PR here, no problem. I didn't know if it was ready to test since @peterfabian left some suggestions on the PR, which I think are valid. Would you like us to test in the current state and let you know if this solves it for us?

Again, I can work with someone @ Pantheon (@carl-alberto?) to get you a testing environment as well. If this needs to happen off of GitHub, just let me know and I will try and get you hooked up with the right folks.

@wss-chadical - YES! But tiny clarification here: Redis doesn't _have_ to be enabled for someone to experience this problem since transients will go into the DB if Redis/another solution isn't present.

wss-chadical commented:
@kloon Happy to test here as well - though ideally not this PR - it's a hot mess.
The original PR is broken. There are multiple work-arounds here as well. One of which we are using - but it's mainly a hack.

Looking forward to a real fix with some lasting power.

ObliviousHarmony committed:
Merge branch 'master' into refactor/23783

ObliviousHarmony commented on includes/wc-core-functions.php:
```diff
@@ -185,7 +185,16 @@ function wc_get_template_part( $slug, $name = '' ) {
+		// Remove ABSPATH from template cache to avoid issues with multi container having different paths.
+		if ( 0 === strpos( $template, ABSPATH ) ) {
+			$template = str_replace( ABSPATH, '', $template );
+			wp_cache_set( $cache_key, $template, 'woocommerce' );
```
You're right about the placement of the cache set, but we need to use ABSPATH here. Since the template might be in a theme, we will just resolve from the installation root.

ObliviousHarmony committed:
Cleaned up the template path replacement and fixed the wc_cache_set positioning

wss-chadical commented:
Testing a preview release with Pantheon right now... they have reworked their container infrastructure so that the /srv/bindings/asdfasdfasdf is now just / - which basically fixes this particular issue.

ObliviousHarmony commented:
I've fixed the issues with the PR as well as added a unit test to ensure that the replacement works as expected. Everything is working as expected for me, but it would be nice to have someone test it in a distributed environment to make sure it works correctly in reality!

JulienMelissas commented:
@wss-chadical - are they offering environments to others to test in?? I need to get on that list, haha.

@ObliviousHarmony - thanks much for your work on this! I will see if we can do some testing next week!

JulienMelissas commented on tests/unit-tests/util/class-wc-tests-core-functions.php:
```diff
@@ -604,6 +604,28 @@ public function test_wc_get_template() {
+		}
+		ob_end_clean();
+
+		remove_filter( 'woocommerce_locatsdae_template', array( $this, 'force_template_path' ) );
```
I think there's a small typo here in `woocommerce_locatsdae_template`? Should be `woocommerce_locate_template` yeah?

ObliviousHarmony replied:
Considering I copy/pasted the add_filter and renamed, I'm not entirely sure how that happened 😢 

Thanks for the catch!

ObliviousHarmony committed:
Corrected a misspelled filter name

wss-chadical commented:
@JulienMelissas open a ticket with support and say someone told you to “Ask support if you can preview project ‘COS’ “

I've been working with support this afternoon - we worked out some kinks. Definitely no more /srv/bindings/asdfasdf/code - it's just /code in all the containers.

Keep in mind we are not in production with pantheon quite yet - tminus 30 days.... so having them muck around in our environments has not been disruptive.

I do not know what else project COS delivers - but for now it is fixing some serious issues in live env we were having.

Good Luck!

Uriel1339 commented:
I can give it a run Wednesday-Friday on my Pantheon installation.

How do I implement the current proposed solution in my installation?

ObliviousHarmony commented:
Here is a version of WooCommerce 3.9.1 that includes only the fixes in this PR:

[woocommerce.zip](https://github.com/woocommerce/woocommerce/files/4155934/woocommerce.zip)

@Uriel1339

Uriel1339 commented:
Tried the build and it works. Thanks @ObliviousHarmony and everyone else involved!

Can't wait to see this in production.

frederickalcantara commented:
I agree with Uriel1339, the newest build works with my Pantheon set up as well. Thanks to the team for updating and making it work with my Pantheon hosted website.

Solarbotics commented:
Any update if this is still the issue for v4.0.x?

vedanshujain committed:
Also cache with WC_ABSPATH to account for when WC is out of ABSPATH.

We already substitute ABSPATH with {{ABSPATH}} token to make sure that exact template path is not cached to support deployment with multiple servers.

This patch also add tokenizing WC_ABSPATH to account for when WooCommerce is installed outside of ABSPATH.

vedanshujain reviewed (dismissed):
@ObliviousHarmony Changes looks good! I have added one more commit to account for cases where WooCommerce is installed outside of ABSPATH. Can you review and approve https://github.com/woocommerce/woocommerce/pull/23971/commits/8f3d8f0495bbbc9bbad6e5dc132f2e1c03354965 as well. I believe we are good merge then.

ObliviousHarmony commented:
> @ObliviousHarmony Changes looks good! I have added one more commit to account for cases where WooCommerce is installed outside of ABSPATH. Can you review and approve [8f3d8f0](https://github.com/woocommerce/woocommerce/commit/8f3d8f0495bbbc9bbad6e5dc132f2e1c03354965) as well. I believe we are good merge then.

I chatted with @claudiosanches last week and he pointed out that you can also relocate the content and plugins folders. I'm going to extract this tokenization logic a little and work on a performant way to remove all of the possible paths from it in a nice sane way.

vedanshujain commented:
Cool. I am dismissing the review then, feel free to ping me again when changes are done.

ObliviousHarmony committed:
Merge branch 'master' into refactor/23783

ObliviousHarmony committed:
Extracted the template path tokenization so that it can be applied to more than one define more readily

ObliviousHarmony committed:
Added tests for path tokenization/untokenization

ObliviousHarmony committed:
Made the path tokenization deterministic

Since the tokens are replaced in a first-discovered first-replaced order, we may accidentally create tokenized paths like '{{ABSPATH}}/test' instead of the desired '{{WP_CONTENT_DIR}}test'. By ordering them according to specificity however, we ensure that we tokenize as much of the path as possible.

ObliviousHarmony committed:
Added a function to fetch all of the path define tokens that may be present in template paths

ObliviousHarmony committed:
Added WPMU_PLUGIN_DIR and removed the path define token for ABSPATH

There's no need for the ABSPATH define since the WC plugin is always a child of a plugin directory anyway.

ObliviousHarmony commented on includes/wc-core-functions.php:
```diff
@@ -150,6 +150,86 @@ function wc_update_order( $args ) {
+		}
+	}
+
+	return apply_filters( 'wc_get_path_define_tokens', $path_tokens );
```
I went with the `wc_` prefix because it was consistent with all of the other template return overrides, which simply share the name of the function. Do you think that consistency might be more important than convention here?

claudiosanches replied:
There's no consistent using `wc_` for filter names, in fact any filter or hook using `wc_` is incorrect and should get deprecated or removed in next versions.
But for some reasons we keep accepting 3rd party Pull Request with incorrect prefixes.

claudiosanches replied:
We have it also documented on: https://github.com/woocommerce/woocommerce/wiki/Naming-conventions

ObliviousHarmony replied:
Sorry, I was referring to `wc_get_template` and `wc_get_template_part`, but I do prefer the more correct prefix :)

ObliviousHarmony committed:
Corrected the filter prefix for get_path_define_tokens

claudiosanches reviewed (approved):
Looks great now!

claudiosanches committed:
Update since tags

claudiosanches reviewed (dismissed):
Need to fix conflicts.

claudiosanches commented:
Hello @carl-alberto @JulienMelissas @Uriel1339 @wss-chadical
Sorry to ping you all here, but I would like to ask if you can test it again, since we made some changes to handle different kinds of WordPress installation, we are looking forward to include this fix on WooCommerce 4.2.0 that will be released in the first week of June.

You can download 4.1.0 with this patch applied in [woocommerce.zip](https://github.com/woocommerce/woocommerce/files/4593958/woocommerce.zip)
Thank you very much for all your help.

carl-alberto commented:
Thanks for working on this @claudiosanches, let us try this out and report back our findings

JulienMelissas commented:
Thanks everyone for working on this. I am still trying to get to a good place with work where I can test this out (Pantheon multidev doesn't have multiple servers so I have to get this into our test environment pipeline which is a bit full of other things we are testing right now). I will report back as soon as I am able to test.

Uriel1339 commented:
Just as an update from my side, this whole Covid situation has us in a
tight spot, so we are going to see to testing as soon as possible, might be
a week or two until we have the opportunity to.

On Tue, May 12, 2020 at 9:35 PM Julien Melissas <notifications@github.com>
wrote:

> Thanks everyone for working on this. I am still trying to get to a good
> place with work where I can test this out (Pantheon multidev doesn't have
> multiple servers so I have to get this into our test environment pipeline
> which is a bit full of other things we are testing right now). I will
> report back as soon as I am able to test.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/woocommerce/woocommerce/pull/23971#issuecomment-627692333>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADFMIKA5XXOUD4JAWKTHAETRRH2NFANCNFSM4HZSVJYA>
> .
>

vanessasickles commented:
A client of ours on Pantheon was having issues due to the issue this patch is supposed to fix.

I can confirm that the patch did fix the issue we were having with the fatal errors and loading the correct templates on both Test and Live environments (for now). I will update if anything else goes wrong or if users start reporting persisting issues.

This patch and this pull request was a life-saver either way -- thank you all!

peterfabian commented:
Before this gets merged, it would be nice if someone from systems (cc @bazza) and VIP (cc @nickdaugherty) could give this a quick look to ensure we won't be breaking anything on those platforms.

peterfabian committed:
Moved unit tests to the new location.

peterfabian committed:
Merge branch 'master' into refactor/23783

claudiosanches commented:
@vanessasickles Thank you very much for testing and letting us know about it, I'm very glad to know that it worked well for you and your customers.

claudiosanches committed:
Update since tag to 4.3.0

vanessasickles commented:
Checking back in! Again, thank you, @claudiosanches, for your work on this.

The patch works fine on its own, but while resolving some issue with our site's core files, we were told by Pantheon support to download the following Pantheon-approved version of WordPress core 5.4.1. https://github.com/pantheon-systems/WordPress/releases/tag/5.4.1

This version of WP, combined with the patched version of WordPress above, causes a similar error to resurface once more. Worked great on our Dev (no load-balancing or server bindings), but every other refresh or so, it throws the error on Live (has multiple servers; same issue as previously had pre-patch).

Here's an attached screenshot of the error. Blocked-out bit is the theme name.
<img width="1181" alt="Screen Shot 2020-06-01 at 3 52 00 PM" src="https://user-images.githubusercontent.com/29470614/83453473-2e3ed580-a420-11ea-9fcc-0b451af9d300.png">

Pantheon support unfortunately seems to not have a solution; they recommended I update you all here.

wss-chadical commented:
@vanessasickles 
Talk to pantheon support or your account manager and insist that your account get placed on the 'COS Preview'.
This problem goes away because all app servers have the same exact file path structures. 
We've been on COS Preview since February.

The COS Preview fixes a multitude of other issues related to file paths and plugins as well. We have found that aside from the this particular issue - many plugins store full system file paths in transients, and general wp config. Creates all kinds of hurt as different app servers grab transients and stored wp config that returns file paths from OTHER app servers.

JulienMelissas commented:
Sorry to derail this conversation but I know of no other way to get a good answer on this: @wss-chadical - we have been asking about the COS preview for months, but getting no information on if it is safe to use in production. Are you on the COS Preview in production with a WooCommerce store that sees any kind of real traffic/sales? Thanks for any insight you can provide.

ObliviousHarmony commented:
Hello @vanessasickles,

This is likely be due to the cache lifetime being set to _forever_ on these paths. This won't be a problem when WooCommerce 4.3.0 releases, as it will bust the cache by changing the version. You _could_ test this theory by incrementing the `$version` property in `includes/class-woocommerce.php`, but I'm not sure what consequences this might have when you actually update and wouldn't advise it on a production server. You could also dump the current cache entries, but I couldn't advise on how to do that with Pantheon.

wss-chadical commented:
@JulienMelissas @vanessasickles 
It's stable. We are on Pantheon Elite, we've been in production for 2 months now. 400K+ visits per month. We have a complex customized wp code base as well. There were some kinks at the beginning (pre go-live) - other than that it has been quite stable. Pantheon will tell you "it's beta, etc" - You can go back to legacy if need be. Does not require any code changes - they just redeploy your appservers...

https://www.wholesalesolar.com

Uriel1339 commented:
On our Pantheon Elite plan we swapped over to this update in WooCommerce 4.2 and removed @JulienMelissas prior hotfix. Everything seems to run fine!

JulienMelissas commented:
For those wondering, we switched to the new COS environments w/ our Pantheon Elite account which solved this problem for us even though we are not able to upgrade to WC 4.2 quite yet. Thanks for the information on that @wss-chadical :)
